manuscripts2/new_functions.py

```diff
@@ -0,0 +1,510 @@
+import logging
+
+from collections import OrderedDict, defaultdict
+from datetime import timezone
+from dateutil import parser
+
+import pandas as pd
+from elasticsearch import Elasticsearch
+from elasticsearch_dsl import A, Q, Search
+
+logger = logging.getLogger(__name__)
+
+
+class Index():
+    """Index class which will represent an index in elasticsearch"""
+
+    def __init__(self, index, url=None):
+        """
+        :url: the address/url of the elasticsearch instance to be used?
+              default: http://localhost:9200/ (optional)
+        :index: name of the elasticsearch index that is to be queried (required)
+        """
+
+        self.url = url if url else None
+        self.index = index  # index to query data from
+        if self.url:
+            self.es = Elasticsearch(self.url)  # This allows us to query _any_ elasticsearch instance
+        else:
+            logger.debug("No custom url provided, using default: http://localhost:9200/")
+            self.es = Elasticsearch("http://localhost:9200/")
+
+
+class Query():
+    """
+    The base class of all the necessary metrics that are going to be calculated.
+    """
+
+    def __init__(self, index_obj, esfilters={}, interval=None, offset=None):
+        """Initialization of necessary parameters
+
+        :index_obj: An Index object containing the connection details
+        :esfilters: TODO: this is still to be implemented
+        :interval: interval to use for timeseries data
+        :offset: TODO: this is still to be implemented
+        """
+        self.es = index_obj.es
+        self.index = index_obj.index
+
+        self.search = Search(using=self.es, index=self.index)  # create the search object
```

Comment:
My main concern with this implementation is related to how you use this search variable, and the others related to it, such as `filters`, `queries`, etc. Why not just use a Search variable, which is extended with anything new you may need when chaining functions?

For example, I would implement `add_query` as:

```
def add_query(self, key_val={}):
    q = Q("match", **key_val)
    self.search = self.search.query(q)
    return self
```

This way, you could use it as:

q = Query().add_query(...)....

(I'm not talking about the parameters for `add_query` here, but about how to implement it).